### PR TITLE
refactor(recsys): share modules and split metrics

### PR DESCRIPTION
**Pull request by haru-256**
## Summary
推薦系 project 間で再利用できる model building block と metrics utility を整理し、project 固有の training flow / model composition は各 project 側に残したまま責務境界を明確にします。

## Related Issues
N/A

## 背景・目的
`projects/recsys-ranking` と `projects/recsys-candidate-generation` は、実験・学習コードを project ごとに持ちつつ、再利用可能な部品は `libs/ml_sandbox_libs` に寄せる方針です。

今回、ranking 側に閉じていた feature interaction / DCNv2 cross-net は、dataset や training flow に依存しない汎用 `nn.Module` building block だったため、shared library 側へ移動しました。一方で、LightningModule wrapper や training/validation step は可読性を優先し、共通化していません。

また、`ml_sandbox_libs.utils.metrics` は入力生成・retrieval metrics・formatting が1ファイルに混在していたため、既存 import 互換を維持しながら `utils/metrics/` package に分割しました。

## Changes
- `FactorizationMachine`, `FirstOrderInteraction`, `SecondOrderInteraction`, `CrossNetV2`, `CrossNetV2MoE` を `libs/ml_sandbox_libs.models.modules` に移動し、unit test も libs 側へ移動
- `ml_sandbox_libs.utils.metrics` を `utils/metrics/` package に分割し、既存の `from ml_sandbox_libs.utils.metrics import ...` は re-export で維持
- recsys model factory の enum 解決 helper を追加し、factory の可読性を改善
- candidate-generation の `cast(...)` による datamodule narrowing を runtime check 付き helper に置換
- ranking の loss factory を `create_loss` から `create_score_loss` に rename し、用途を明確化
- project-local に残っていた旧 module/test directory を cleanup

## How to Test
- [x] `cd libs/ml_sandbox_libs && make lint`
- [x] `cd libs/ml_sandbox_libs && uv run pytest tests/test_models/test_modules/test_interaction.py tests/test_models/test_modules/test_cross_net.py tests/test_utils/test_metrics.py -v` (72 passed)
- [x] `cd projects/recsys-ranking && make lint && make test` (96 passed)
- [x] `cd projects/recsys-candidate-generation && make lint && make test` (48 passed)

## Additional Context
`cd libs/ml_sandbox_libs && make test` は、`test_bipartite_graph_transform_preserves_public_indices_for_loader_batches` で `NeighborSampler` が `pyg-lib` または `torch-sparse` を要求するため 1 件失敗します。今回の refactor とは無関係な環境依存 failure と判断しています。

### CODE REVIEW

**Comment by haru-256** [comment]
PR指摘とCI failure に対応しました。

- `metrics.__all__` を ruff RUF022 に合わせて並び替え
- `create_retrieval_inputs` の `indexes` を `score.device` 上で作成するよう修正
- device invariant の test assertion を追加
- `create_score_loss` の `cfg` docstring を full config 前提に修正

確認:
- `cd libs/ml_sandbox_libs && uv run ruff check src/ml_sandbox_libs/utils/metrics tests/test_utils/test_metrics.py`
- `cd libs/ml_sandbox_libs && uv run mypy src/ml_sandbox_libs/utils/metrics tests/test_utils/test_metrics.py`
- `cd libs/ml_sandbox_libs && uv run pytest tests/test_utils/test_metrics.py -v` (13 passed)
- `cd projects/recsys-ranking && uv run ruff check src/loss/factory.py`
- `cd projects/recsys-ranking && uv run mypy src/loss/factory.py`
- `cd projects/recsys-ranking && uv run pytest src/tests/test_loss.py -v` (7 passed)

File: libs/ml_sandbox_libs/src/ml_sandbox_libs/utils/metrics/inputs.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    Returns:
+        score: logits, shape (batch_size, pos_sample_size + neg_sample_size)
+        target: target long, shape (batch_size, pos_sample_size + neg_sample_size)
+        indexes: indexes, shape (batch_size, pos_sample_size + neg_sample_size)
+    """
+    score = torch.cat([positive, negative], dim=1)
+    target = torch.cat([torch.ones_like(positive), torch.zeros_like(negative)], dim=1).long()
+    batch_size, num_samples = score.size()
+    indexes = torch.arange(batch_size, device=score.device).reshape(batch_size, 1).expand(batch_size, num_samples).long()
```
[the remark this answers is not in the document]
Comment (haru-256) [reply libs/ml_sandbox_libs/src/ml_sandbox_libs/utils/metrics/inputs.py]: 対応しました。`indexes` 生成を `unsqueeze(1).expand(...)` に変更し、冗長な `.long()` を削除しました。`device=score.device` は維持しているため、前回修正した device placement は変わりません。

確認:
- `cd libs/ml_sandbox_libs && uv run ruff check src/ml_sandbox_libs/utils/metrics/inputs.py tests/test_utils/test_metrics.py`
- `cd libs/ml_sandbox_libs && uv run mypy src/ml_sandbox_libs/utils/metrics/inputs.py tests/test_utils/test_metrics.py`
- `cd libs/ml_sandbox_libs && uv run pytest tests/test_utils/test_metrics.py -v` (13 passed)
